JWT Bearer token support

tilgovi opened the pull request:
Change the authorization code to normalize the annotator auth tokens
and headers into standard JWT bearer tokens and the auth token header
into the HTTP Authorization header.

Start accepting JWT Bearer tokens as authorization grants. When used
in this manner, the issuing client must already be authorized for the
subject. Currently, the only valid issuer is the built-in web client,
which is always pre-authorized. In the future, it will be necessary to
store validations in a database and only allow the JWT Bearer Token
flow generate tokens for users that have already completed another
flow with the client that allows for the use of refresh tokens. In
essence, the JWT Bearer Token is a self-issued, pre-authorized refresh
token when used as an authorization grant.

aron commented on h/authentication.py:
```diff
@@ -21,6 +21,13 @@
 
+def normalize_authorization_header(request):  # bw compat
+    """Convert any X-Annotator-Auth-Token header to an Authorization header."""
+    token = request.headers.pop('X-Annotator-Auth-Token', None)
```
Do we actually want to remove the header here or just read it? I'm just wondering about the value of other pieces of code needing to know if the original header was present. Usually in these cases, a new normalized header is added and used in the code, but the original is left.

tilgovi replied:
I don't care that much either way but my thought was that this helped ensure that no other code anywhere in the back-end mentioned or depended on this header anymore.

aron commented on h/authorization.py:
```diff
@@ -32,16 +32,32 @@
+EPOCH = datetime(1970, 1, 1).replace(tzinfo=iso8601.iso8601.UTC)
+
+
+def now():
```
Consider calling this `utc_now()` to reinforce that the timezone is different from just calling `datetime.now()`?

tilgovi replied:
:+1:

aron commented on h/authorization.py:
```diff
@@ -66,6 +82,48 @@ def is_web_client(request, client_id):
+        token['iat'] = posix_seconds(iso8601.parse_date(token['issuedAt']))
+        if 'ttl' in token:
+            token['exp'] = token['iat'] + int(token['ttl'])
+            del token['ttl']
```
I'd prefer we cloned the `token` object and returned the original instance. Otherwise the caller needs to be aware that the token argument will be modified.

tilgovi replied:
Fine with that. I had it that way to begin with, but changed it to match the behavior of the normalize function for the header. Heh.

aron commented on h/authorization.py:
```diff
@@ -113,15 +175,48 @@ def save_bearer_token(self, token, request):
+            # TODO: client certificates instead of secrets
+            jwt.decode(token, client.client_secret)
+        except jwt.DecodeError:
+            return False
```
This is now a seriously long method, how about we break it up into smaller functions that do each check.

tilgovi replied:
It's long-ish, but it's also code that's not used anywhere else, is a series of very simple conditionals, and I'd still want to only expose the same public API and therefore test the same interface, so I don't know that it helps much to break it up.

tilgovi replied:
As you like.

aron commented on h/authorization.py:
```diff
@@ -66,6 +82,48 @@ def is_web_client(request, client_id):
+        del token['userId']
+
+
+def token_generator(request):  # bw compat
```
Perhaps rename this to `annotator_token_generator` to distinguish it from the JWT bearer token?

tilgovi replied:
This is *the* token generator. With this PR we write old tokens and read both. We either need to be able to disambiguate here between which client we're doing this for or we need to do the migration client-side by having a generation that supports the new format before we roll it out on the backend, wholesale replace this function here, and then remove the deprecated front end code that deals with the old format.

aron commented on h/oauth/grant_types/jwt_bearer.py:
```diff
@@ -0,0 +1,53 @@
+                raise errors.InvalidClientError(request=request)
+            else:
+                request.client_id = request.client_id or \
+                    request.client.client_id
```
Why is this re-assigning the `request.client_id`? Is it possible for the value to be changed between here and line 30?

tilgovi replied:
Line 30 it may be `client_id` or `client_assertion` that triggers client authentication. After authentication, `request.client` should be set, so we ensure at that point we copy the `request.client.client_id`. This mirrors code from oauthlib near exactly, with the thought that I would like to break it out and submit it as a patch there.

aron commented:
I've just run through this and made a few small comments, if others could review my comments and +/-1 them I'll fix them so we don't have to wait for @tilgovi in order to merge this.

I'd also appreciate a couple more reviews on the changes as it's been a while since I last implemented an oauth flow.

tilgovi commented:
Some quick replies. Yes, I know I'm on vacation. It is glorious. I am just poking my head in while I check around for flights to my next destination.

jtremback commented:
Come to Berlin! I'm leaving on the 27th or 28th.

tilgovi commented:
@jtremback sorry! I'm at the airport heading to Barcelona. :-) see you back home.

tilgovi commented:
I think I've addressed your concerns, @aron:

- I've started generating tokens that have all the claims in them, standard and Annotator-style. This should satisfy old clients while avoiding the need to validate old tokens (a very small window of incompatibility will occur for sidebars that are not reloaded, but I think we should wave this away at this stage).
- I upgraded to a newer version of PyJWT which has a nice public API that separates parsing the token from verifying the token. This means we can avoid re-implementing the validation logic, since it handles the standard claims, and validate_bearer_token becomes much shorter and simpler.
- I'm going to leave the JWTBearerGrant function as is. It's rather long but it follows the style of OAuthLib very closely. A lot of the length comes from the formatting of the code (long lines) and logging statements. I'm reluctant to change this.

Thanks for the review. I'm very happy with this PR at this point. @nickstenning, @gergely-ujvari, @csillag if any of you could have a look, that'd be awesome.

aron commented:
:thumbsup:

nickstenning commented on h/authorization.py:
```diff
@@ -32,40 +32,62 @@
+    ttl = datetime.timedelta(seconds=DEFAULT_TTL)
+
+    payload = {
+        'aud': request.host_url,
```
I know what these fields mean, but only because I've worked with JWTs before. It might be worth adding comments to explain them, or at least a link to the relevant part of the JWT RFC.

nickstenning commented:
Overall this is looking great, but I don't see any tests for `h.oauth.JWTBearerGrant` :)

tilgovi commented:
Thanks for the reviews. I found some things I wanted to change and I refactored this. I'm closing it and will open a series of smaller PRs.
